### PR TITLE
Refactor/receipt api

### PR DESCRIPTION
**Pull request by ravinderk**
## Description

As pr @JasonTheAdams Comment: https://github.com/impress-org/give-form-field-manager/pull/312#pullrequestreview-412044887

We find out that current API has the flexibility to register new detail group and item but it is not flexible enough to accept a custom object that means we accept a class name for detail group and item and create object internally but in case of FFM: https://github.com/impress-org/give-form-field-manager/pull/312 we find out that we need the flexibility to register custom object creation method. This pr will resolve this issue.
 
## Affects

Receipt API related files. Removed existing filter and added new action hooks.

## What to test

Tested this pr against https://github.com/impress-org/give-form-field-manager/pull/312

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

### CODE REVIEW

File: src/Form/Template.php [hunk 1]
```diff
@@ -256,6 +256,15 @@ public function getDonationIntroductionContentPosition() {
...
 	public function getReceiptDetails( $donationId ) {
-		return new Receipt( $donationId );
+		$receipt = new Receipt( $donationId );
+
+		/**
+		 * fire the action for receipt object.
+		 *
+		 * @since 2.7.0
+		 */
+		do_action( 'give_new_receipt', $receipt );
```
Comment (JasonTheAdams) [inline src/Form/Template.php:266]: Why is the receipt being constructed in two different places? Here and in the `Sequoia` class.

File: src/Receipt/Receipt.php [hunk 2]
```diff
@@ -88,35 +71,61 @@ private function setupReceipt() {
...
+		$object = new $detailGroupClassName( $this->donationId );
+
 		/**
-		 * Use this filter to handle object initialization for your detail group class.
+		 * fire the action for receipt detail group.
 		 *
 		 * @since 2.7.0
 		 */
-		$object = apply_filters( 'give_receipt_create_detail_group_object', null, $class, $this->donationId );
+		 do_action( 'give_new_receipt_detail_group', $object );
```
Comment (JasonTheAdams) [inline src/Receipt/Receipt.php:84]: So is the idea, here, that a `DetailGroup` subclass isn't constructed until it's retrieved? And is this something an addon will be using?

File: src/Form/Template.php (hunk 1 above)
Comment (ravinderk) [reply src/Form/Template.php:266]: @JasonTheAdams This is an abstract class, so I added a default function. Any template have ability to overwrite it.

File: src/Receipt/Receipt.php (hunk 2 above)
Comment (ravinderk) [reply src/Receipt/Receipt.php:84]: @JasonTheAdams by using this hook and `DetailGroup ` subclass object they can easily add a new detail item to the group as I am using it in FFM pr: https://github.com/impress-org/give-form-field-manager/pull/312

**Comment by ravinderk** [comment]
> Left some comments, @ravinderk. You're clearly doing some fancy stuff here with subclass the groups. This PR doesn't give me a full picture, but perhaps I don't need one. In my mind the Receipt, ReceiptGroup, and such were purely data driven — that is, simple objects which just held the data to be displayed in the receipt view. But it seems like they're also responsible for retrieving the data to be displayed?

@JasonTheAdams Receipt object only responsible for providing data. it is form template responsibility to render it how they want for like we are doing in `Sequoia/view/receipt.php`

**Comment by JasonTheAdams** [comment]
> @JasonTheAdams Receipt object only responsible for providing data. it is form template responsibility to render it how they want for like we are doing in `Sequoia/view/receipt.php`
@ravinderk I guess my question is, then, where are you binding the donation data to the receipt?

**Comment by ravinderk** [comment]
@JasonTheAdams I updated Receipt API and removed complexity.
As per new logic `Receipt` is a `abstract` class and `DonationReceipt` a class to get donation receipt. We are accepting array from functions `addSection` and `AddLineItem` and return `stdClass` object when call `$receipt->getSections()`.

Section Object:
```
stdClass Object
(
    [Donor] => stdClass Object
        (
            [id] => Donor
            [label] => Donation Details
            [lineItems] => stdClass Object
                (
                    [fullName] => stdClass Object
                        (
                            [icon] => <i class="fas fa-user"></i>
                            [id] => fullName
                            [label] => Donor Name
                            [value] => Ravinder Kumar
                        )

                    [emailAddress] => stdClass Object
                        (
                            [icon] => <i class="fas fa-envelope"></i>
                            [id] => emailAddress
                            [label] => Email Address
                            [value] => ravinder.kumar.apj@gmail.com
                        )

                    [billingAddress] => stdClass Object
                        (
                            [icon] => <i class="fas fa-envelope"></i>
                            [id] => billingAddress
                            [label] => Billing Address
                            [value] => <br>,<br>
                        )

                )

        )

    [Donation] => stdClass Object
        (
            [id] => Donation
            [lineItems] => stdClass Object
                (
                    [paymentMethod] => stdClass Object
                        (
                            [icon] => 
                            [id] => paymentMethod
                            [label] => Payment Method
                            [value] => Test Donation
                        )

                    [paymentStatus] => stdClass Object
                        (
                            [icon] => 
                            [id] => paymentStatus
                            [label] => Payment Status
                            [value] => Complete
                        )

                    [amount] => stdClass Object
                        (
                            [icon] => 
                            [id] => amount
                            [label] => Donation Amount
                            [value] => $50
                        )

                    [totalAmount] => stdClass Object
                        (
                            [icon] => 
                            [id] => totalAmount
                            [label] => Donation Total
                            [value] => $50
                        )

                )

        )

    [AdditionalInformation] => stdClass Object
        (
            [id] => AdditionalInformation
            [label] => Additional Information
        )

)
```

Let me know what do you think.

**Comment by JasonTheAdams** [comment]
@ravinderk Wait, so did you get rid of the group and line-item classes entirely are just using `stdClass` now for everything?

**Comment by ravinderk** [comment]
@JasonTheAdams yes.

**Comment by ravinderk** [comment]
@JasonTheAdams This pr ready to review. I updated pr as per your suggestion.
`DonationReceipt` class extends `Receipt` class. Receipt made of sections `[Section class]` and section made of line items`[LineItem class]`. `DonationReceipt` and `Section` has `public` functions to add and remove items. Because  `DonationReceipt` and `Section` implements `Iterator`, so it is easy to loop over an object as an array and implements `ArrayAccess`  allow to perform array related operation on an object.


```
Array
(
    [Donor] => Give\Receipt\Section Object
        (
            [label] => Donation Details
            [id] => Donor
            [lineItems:protected] => Array
                (
                    [fullName] => Give\Receipt\LineItem Object
                        (
                            [id] => fullName
                            [label] => Donor Name
                            [value] => Ravinder Kumar
                            [icon] => <i class="fas fa-user"></i>
                        )

                    [emailAddress] => Give\Receipt\LineItem Object
                        (
                            [id] => emailAddress
                            [label] => Email Address
                            [value] => ravinder.kumar.apj@gmail.com
                            [icon] => <i class="fas fa-envelope"></i>
                        )

                )

        )

    [Donation] => Give\Receipt\Section Object
        (
            [label] => 
            [id] => Donation
            [lineItems:protected] => Array
                (
                    [paymentMethod] => Give\Receipt\LineItem Object
                        (
                            [id] => paymentMethod
                            [label] => Payment Method
                            [value] => Test Donation
                            [icon] => 
                        )

                    [paymentStatus] => Give\Receipt\LineItem Object
                        (
                            [id] => paymentStatus
                            [label] => Payment Status
                            [value] => Complete
                            [icon] => 
                        )

                    [amount] => Give\Receipt\LineItem Object
                        (
                            [id] => amount
                            [label] => Donation Amount
                            [value] => $50
                            [icon] => 
                        )

                    [totalAmount] => Give\Receipt\LineItem Object
                        (
                            [id] => totalAmount
                            [label] => Donation Total
                            [value] => $50
                            [icon] => 
                        )

                )

        )

    [AdditionalInformation] => Give\Receipt\Section Object
        (
            [label] => Additional Information
            [id] => AdditionalInformation
            [lineItems:protected] => Array
                (
                )

        )

)

Array
(
    [Donor] => Give\Receipt\Section Object
        (
            [label] => Donation Details
            [id] => Donor
            [lineItems:protected] => Array
                (
                    [fullName] => Give\Receipt\LineItem Object
                        (
                            [id] => fullName
                            [label] => Donor Name
                            [value] => Ravinder Kumar
                            [icon] => <i class="fas fa-user"></i>
                        )

                    [emailAddress] => Give\Receipt\LineItem Object
                        (
                            [id] => emailAddress
                            [label] => Email Address
                            [value] => ravinder.kumar.apj@gmail.com
                            [icon] => <i class="fas fa-envelope"></i>
                        )

                )

        )

    [Donation] => Give\Receipt\Section Object
        (
            [label] => 
            [id] => Donation
            [lineItems:protected] => Array
                (
                    [paymentMethod] => Give\Receipt\LineItem Object
                        (
                            [id] => paymentMethod
                            [label] => Payment Method
                            [value] => Test Donation
                            [icon] => 
                        )

                    [paymentStatus] => Give\Receipt\LineItem Object
                        (
                            [id] => paymentStatus
                            [label] => Payment Status
                            [value] => Complete
                            [icon] => 
                        )

                    [amount] => Give\Receipt\LineItem Object
                        (
                            [id] => amount
                            [label] => Donation Amount
                            [value] => $50
                            [icon] => 
                        )

                    [totalAmount] => Give\Receipt\LineItem Object
                        (
                            [id] => totalAmount
                            [label] => Donation Total
                            [value] => $50
                            [icon] => 
                        )

                )

        )

    [AdditionalInformation] => Give\Receipt\Section Object
        (
            [label] => Additional Information
            [id] => AdditionalInformation
            [lineItems:protected] => Array
                (
                )

        )

)


```

File: src/Receipt/DonationReceipt.php [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+	 *
+	 * @param  array  $section
+	 * @param  string $position Position can be set either "before" or "after" to insert section at specific position.
+	 * @param  string $sectionId
+	 *
+	 * @return Section
+	 *
+	 * @since 2.7.0
+	 */
+	public function addSection( $section, $position = '', $sectionId = '' ) {
```
Comment (JasonTheAdams) [inline src/Receipt/DonationReceipt.php:76]: This isn't specific to donations, so you can just move it into the parent `Receipt` class. Same thing with the validation class.

File: src/Receipt/Receipt.php [hunk 4]
```diff
@@ -1,21 +1,20 @@
...
  * This class represent receipt as object.
- * Receipt can have multiple detail group and detail group can has multiple detail item.
- * You can add your own logic to render receipt because it does not return any UI item.
+ * Receipt can have multiple sections and sections can have multiple line items.
  *
  * @since 2.7.0
  * @package Give\Receipt
  */
-class Receipt {
+abstract class Receipt implements Iterator, ArrayAccess {
```
Comment (JasonTheAdams) [inline src/Receipt/Receipt.php:17]: Go ahead and move the `Iterator` and `ArrayAccess` methods into this abstract class. They're not specific to donations.

File: src/Receipt/Receipt.php [hunk 5]
```diff
@@ -32,91 +31,39 @@ class Receipt {
...
-	 * @since 2.7.0
+	 * @param  string $sectionId
 	 *
-	 * @return array
+	 * @since 2.7.0
 	 */
-	public function getDetailGroupList() {
-		return $this->detailsGroupList;
-	}
+	abstract public function removeSection( $sectionId );
```
Comment (JasonTheAdams) [inline src/Receipt/Receipt.php:68]: Just define the method in here, no need to force the child class to define it.

Comment (ravinderk) [reply src/Receipt/Receipt.php:68]: Make sense.

File: src/Receipt/DonationReceipt.php (hunk 3 above)
Comment (ravinderk) [reply src/Receipt/DonationReceipt.php]: fixed: https://github.com/impress-org/givewp/compare/ccb59a12b626...55c57984a9f5

File: src/Receipt/Receipt.php (hunk 4 above)
Comment (ravinderk) [reply src/Receipt/Receipt.php:17]: fixed: https://github.com/impress-org/givewp/compare/ccb59a12b626...55c57984a9f5

File: src/Receipt/Receipt.php (hunk 5 above)
Comment (ravinderk) [reply src/Receipt/Receipt.php]: fixed: https://github.com/impress-org/givewp/compare/ccb59a12b626...55c57984a9f5

**Comment by ravinderk** [comment]
@JasonTheAdams I implemented your suggestions. I am merging this pr and let me know if other suggestions or any quest about the recent updates.

**Pull request merged**